board/drivers/fdcan.h

```diff
@@ -0,0 +1,195 @@
+  uint8_t bus_number = BUS_NUM_FROM_CAN_NUM(can_number);
+	uint8_t rx_fifo_idx;
+
+	// Rx FIFO 0 new message
+	if((CANx->IR & FDCAN_IR_RF0N) != 0) {
+    CANx->IR |= FDCAN_IR_RF0N;
+    while((CANx->RXF0S & FDCAN_RXF0S_F0FL) != 0) {
+      can_rx_cnt += 1;
+
+      // can is live
+      pending_can_live = 1;
+
+      // getting new message index (0 to 63)
+      rx_fifo_idx = (uint8_t)((CANx->RXF0S >> FDCAN_RXF0S_F0GI_Pos) & 0x3F);
+
+      uint32_t RxFIFO0SA = FDCAN_START_ADDRESS + (can_number * FDCAN_OFFSET);
+      CAN_FIFOMailBox_TypeDef to_push;
+      CAN_FIFOMailBox_TypeDef *fifo;
+
+      // getting address
+      fifo = (CAN_FIFOMailBox_TypeDef *)(RxFIFO0SA + (rx_fifo_idx * FDCAN_RX_FIFO_0_EL_SIZE));
+
+      // Need to convert real CAN frame format to mailbox "type"
+      to_push.RIR = ((fifo->RIR >> 27) & 0x4) | ((fifo->RIR >> 28) & 0x2) | (fifo->RIR << 3); // identifier format | frame type | identifier
+      to_push.RDTR = ((fifo->RDTR >> 16) & 0xF) | (fifo->RDTR << 16); // DLC (length) | timestamp
+      to_push.RDLR = fifo->RDLR;
+      to_push.RDHR = fifo->RDHR;
+
+      // modify RDTR for our API
+      to_push.RDTR = (to_push.RDTR & 0xFFFF000F) | (bus_number << 4);
+
+      // forwarding (panda only)
+      int bus_fwd_num = (can_forwarding[bus_number] != -1) ? can_forwarding[bus_number] : safety_fwd_hook(bus_number, &to_push);
+      if (bus_fwd_num != -1) {
+        CAN_FIFOMailBox_TypeDef to_send;
+        to_send.RIR = to_push.RIR;
+        to_send.RDTR = to_push.RDTR;
+        to_send.RDLR = to_push.RDLR;
+        to_send.RDHR = to_push.RDHR;
+        can_send(&to_send, bus_fwd_num, true);
+      }
+
+      can_rx_errs += safety_rx_hook(&to_push) ? 0U : 1U;
+      ignition_can_hook(&to_push);
+
+      current_board->set_led(LED_BLUE, true);
+      can_send_errs += can_push(&can_rx_q, &to_push) ? 0U : 1U;
+
+      // update read index 
+      CANx->RXF0A = rx_fifo_idx;
+    }
+
+  } else if((CANx->IR & (FDCAN_IR_PEA | FDCAN_IR_PED | FDCAN_IR_RF0L | FDCAN_IR_RF0F | FDCAN_IR_EW | FDCAN_IR_MRAF | FDCAN_IR_TOO)) != 0) {
+    #ifdef DEBUG
+      puts("FDCAN error, FDCAN_IR: ");puth(CANx->IR);puts("\n");
+    #endif
+    CANx->IR |= (FDCAN_IR_PEA | FDCAN_IR_PED | FDCAN_IR_RF0L | FDCAN_IR_RF0F | FDCAN_IR_EW | FDCAN_IR_MRAF | FDCAN_IR_TOO); // Clean all error flags
+    can_err_cnt += 1;
+  } else { 
+    
+  }
+  
+}
+
+void FDCAN1_IT0_IRQ_Handler(void) { can_rx(0); }
+void FDCAN1_IT1_IRQ_Handler(void) { process_can(0); }
+
+void FDCAN2_IT0_IRQ_Handler(void) { can_rx(1); }
+void FDCAN2_IT1_IRQ_Handler(void) { process_can(1); }
+
+void FDCAN3_IT0_IRQ_Handler(void) { can_rx(2);  }
+void FDCAN3_IT1_IRQ_Handler(void) { process_can(2); }
+
+bool can_init(uint8_t can_number) {
+  bool ret = false;
+
+  REGISTER_INTERRUPT(FDCAN1_IT0_IRQn, FDCAN1_IT0_IRQ_Handler, CAN_INTERRUPT_RATE, FAULT_INTERRUPT_RATE_CAN_1)
+  REGISTER_INTERRUPT(FDCAN1_IT1_IRQn, FDCAN1_IT1_IRQ_Handler, CAN_INTERRUPT_RATE, FAULT_INTERRUPT_RATE_CAN_1)
```

Comment:
why do they all have different fault rates?

Reply:
I don't know, last time you worked on these rates in PR #580 and they are in cereal also. (And it is not a rate, but fault flag)